### ISSUE TITLE
nvda does not announce hyperlink for JAB applications

### ISSUE THREAD

**Issue by rpaquay** [nvaccess/nvda#5814]
For example, when opening the "JEditorPaneHTML Demo" tab of the SwingSet2 application, the page displayed contains 6 hyperlinks toward the end of the document (`Title Page`, `To The King`, `The Preface`, etc.)

It seems that nvda does not see these words as hyperlinks and there is no way to navigate the links.

This seems specific to JAB applications.

**Comment by rpaquay** [comment]
From an implementation point of view, I think the problem is that nvda does not consume any of the JAB api related to hyperlinks (e.g. `GetAccessibleHypertextExt`, `GetAccessibleHyperlinkCount`, `GetAccessibleHypertextLinkIndex`, `GetAccessibleHyperlink`).

I can see the correct hyperlink data using the Access Bridge Explorer (https://github.com/google/access-bridge-explorer). When selecting the text pane in the "Accessibility Tree", the "Component Properties" tab contains a group called "Accessible Hyper Text" than contains 6 hyperlinks, corresponding the the 6 hyperlinks of the displayed text in SwingSet2.

**Comment by jcsteh** [comment]
It's true that we'd need to implement the hyperlink API. However, there are other potential issues:

* Do these controls actually support caret navigation? If they don't, the user won't be able to move to the links anyway. Also, hey won't be able to navigate the text except using the review cursor. We do support this in other browsers using browse mode, but that would require a browse mode implementation for JAB, which is a fairly big chunk of work.
* Do these controls support interactive HTML elements other than links? If so, they're probably represented by embedded objects, but I don't think JAB will provide us a way to efficiently get to those objects.

**Comment by rpaquay** [comment]
@jcsteh

* Yes, these controls support caret navigation. They are basically simple text editor (JEditorPane in Swing) with the ability to add embed objects in them. The objects are accessible through the list of children from the accessible context of the text editor. The hyperlinks are accessible directly via accessible context of the text editor itself.
* I don't know if other interactive elements types are supported, but I do know it is possible to add more than just text. All the embedded objects are accessible through the list of children, I believe.

**Comment by jcsteh** [comment]
The problem with embedded objects as simply a list of children is that 
if you retrieve a range in the middle of the text, you can't get the 
embedded object. You have to retrieve the text from the start f the 
control every time (even if the user moves by character) so you can 
count the embedded objects. That's horrible from a perf perspective.

Nevertheless, we do have code similar to this for OpenOffice. See 
compoundDocuments.TreeCompoundTextInfo.

The question is how complex this gets. Does it expose every paragraph as 
a separate context? Does it allow multiple levels of hierarchy? Does it 
expose wrapping divs, etc. in the hierarchy or are they flattened out?

**Comment by bhavyashah** [comment]
rpaquay  Would you or someone else be able to respond to the following excerpt of https://github.com/nvaccess/nvda/issues/5814#issuecomment-198106756: "The question is how complex this gets. Does it expose every paragraph as a separate context? Does it allow multiple levels of hierarchy? Does it expose wrapping divs, etc. in the hierarchy or are they flattened out?"

**Comment by rpaquay** [comment]
@bhavyashah I don't recall the context around this. However, my earlier comment (https://github.com/nvaccess/nvda/issues/5814#issuecomment-197459029) seems to suggest Access Bridge Explorer exposes all the information exposed by JEditorPane, so it is a matter of creating a sample app and looking at the accessibility tree.

**Comment by Adriani90** [comment]
cc: @kaleev

**Comment by Adriani90** [comment]
cc: @thgcode since you are familliar with JAB, maybe you have some suggestions here or can contribute to fixing this.
